metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -0,0 +1,900 @@
+        if (partitionInfo == null) {
+            return new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION,
+                "No such partition as " + topic + "-" + partitionId);
+        }
+        int newLeader = chooseNewLeader(partitionInfo, partitionInfo.isr, unclean);
+        if (newLeader == partitionInfo.leader) {
+            if (newLeader < 0) {
+                return new ApiError(Errors.LEADER_NOT_AVAILABLE,
+                    "Unable to find any leader for the partition.");
+            } else {
+                return ApiError.NONE;
+            }
+        } else {
+            int[] newIsr = partitionInfo.isr;
+            if (!Replicas.contains(partitionInfo.isr, newLeader)) {
+                newIsr = new int[] {newLeader};
+            }
+            records.add(new ApiMessageAndVersion(new PartitionChangeRecord().
+                setPartitionId(partitionId).
+                setTopicId(topicId).
+                setIsr(Replicas.toList(newIsr)).
+                setLeader(newLeader).
+                setLeaderEpoch(partitionInfo.leaderEpoch + 1).
+                setPartitionEpoch(partitionInfo.partitionEpoch + 1), (short) 0));
+            return ApiError.NONE;
+        }
+    }
+
+    int chooseNewLeader(PartitionControlInfo partition, int[] newIsr, boolean unclean) {
+        for (int i = 0; i < partition.replicas.length; i++) {
+            int replica = partition.replicas[i];
+            if (Replicas.contains(newIsr, replica)) {
+                return replica;
+            }
+        }
+        if (unclean) {
+            for (int i = 0; i < partition.replicas.length; i++) {
+                int replica = partition.replicas[i];
+                if (clusterControl.unfenced(replica)) {
+                    return replica;
+                }
+            }
+        }
+        return -1;
+    }
+
+    ControllerResult<BrokerHeartbeatReply> processBrokerHeartbeat(
+                BrokerHeartbeatRequestData request, long lastCommittedOffset) {
+        int brokerId = request.brokerId();
+        long brokerEpoch = request.brokerEpoch();
+        clusterControl.checkBrokerEpoch(brokerId, brokerEpoch);
+        BrokerHeartbeatManager heartbeatManager = clusterControl.heartbeatManager();
+        BrokerControlStates states = heartbeatManager.calculateNextBrokerState(brokerId,
+            request, lastCommittedOffset, brokersToIsrs.hasLeaderships(brokerId));
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        if (states.current() != states.next()) {
+            switch (states.next()) {
+                case FENCED:
+                    handleBrokerFenced(brokerId, records);
+                    break;
+                case UNFENCED:
+                    handleBrokerUnfenced(brokerId, brokerEpoch, records);
+                    break;
+                case CONTROLLED_SHUTDOWN:
+                    removeLeaderships(brokerId, records);
```

Comment:
As I mentioned above, I changed it so that it now removes the broker from all non-singleton ISRs, as well as removing it from leaderships.

It seems like the remaining behavioral difference is that the new code will, if no other leader can be chosen, set the leader to -1 (offline).  If we don't do this, controlled shutdown easily gets stuck if there are any partitions with replication factor = 1.  Maybe we can tune this a bit later?

I like the idea of including the partition epoch in the follower fetch request.